File: javascript/packages/linter/src/rules/a11y-nested-interactive-elements.ts
import { ElementStackVisitor } from "./rule-utils.js"
import { getTagLocalName, getStaticAttributeValue } from "@herb-tools/core"

import { ParserRule } from "../types.js"
import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLElementNode, ParseResult, ParserOptions } from "@herb-tools/core"

const INTERACTIVE_ELEMENTS = new Set([
  "a",
  "button",
  "input",
  "select",
  "summary",
  "textarea",
])

class NestedInteractiveElementsVisitor extends ElementStackVisitor {
  visitHTMLElementNode(node: HTMLElementNode): void {
    const tagName = getTagLocalName(node)

    if (tagName && INTERACTIVE_ELEMENTS.has(tagName)) {
      if (tagName === "input" && getStaticAttributeValue(node, "type") === "hidden") {
        super.visitHTMLElementNode(node)
        return
      }

      const ancestor = this.findInteractiveAncestor(tagName)

      if (ancestor) {
        this.addOffense(
          `Found \`<${tagName}>\` nested inside of \`<${ancestor}>\`. Nesting interactive elements produces invalid HTML, and assistive technologies, such as screen readers, might ignore or respond unexpectedly to such nested controls.`,
          node.location,
        )
      }
    }

    super.visitHTMLElementNode(node)
  }

  private findInteractiveAncestor(childTagName: string): string | null {
    const ancestor = this.currentElement

    if (!ancestor) return null

    const ancestorTag = getTagLocalName(ancestor)
    if (!ancestorTag || !INTERACTIVE_ELEMENTS.has(ancestorTag)) return null

    if (ancestorTag === "summary" && childTagName === "a") return null

    return ancestorTag
  }
}

export class A11yNestedInteractiveElementsRule extends ParserRule {
  static ruleName = "a11y-nested-interactive-elements"
  static introducedIn = this.version("unreleased")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: false,
      severity: "error"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      action_view_helpers: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new NestedInteractiveElementsVisitor(this.ruleName, context)
    visitor.visit(result.value)
    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/a11y_nested_interactive_elements.rs
use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::element_stack::ElementStack;
use crate::utils::tag_utils::{get_element_static_attribute_value, get_tag_local_name};

use herb::nodes::HTMLElementNode;
use herb::ParseResult;
use herb::Visitor;
use herb_config::{Severity, SeverityConfig};

const INTERACTIVE_ELEMENTS: &[&str] = &["a", "button", "input", "select", "summary", "textarea"];

pub struct A11yNestedInteractiveElementsRule;

struct NestedInteractiveElementsVisitor {
  rule_name: &'static str,
  offenses: Vec<UnboundOffense>,
  element_stack: ElementStack,
}

impl NestedInteractiveElementsVisitor {
  fn find_interactive_ancestor(&self, child_tag_name: &str) -> Option<String> {
    for ancestor in self.element_stack.ancestors() {
      if !INTERACTIVE_ELEMENTS.contains(&ancestor) {
        continue;
      }

      if ancestor == "summary" && child_tag_name == "a" {
        continue;
      }

      return Some(ancestor.to_string());
    }

    None
  }
}

impl Visitor for NestedInteractiveElementsVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    let tag_name = get_tag_local_name(node);

    let is_hidden_input = tag_name.as_deref() == Some("input") && get_element_static_attribute_value(node, "type").as_deref() == Some("hidden");

    if let Some(ref tag_name) = tag_name {
      if INTERACTIVE_ELEMENTS.contains(&tag_name.as_str()) && !is_hidden_input {
        if let Some(ancestor) = self.find_interactive_ancestor(tag_name) {
          self.offenses.push(UnboundOffense::new(
            self.rule_name,
            format!(
              "Found `<{}>` nested inside of `<{}>`. Nesting interactive elements produces invalid HTML, and assistive technologies, such as screen readers, might ignore or respond unexpectedly to such nested controls.",
              tag_name, ancestor
            ),
            node.location.clone(),
          ));
        }
      }
    }

    self.element_stack.push_optional(tag_name);
    self.walk_html_element_node(node);
    self.element_stack.pop();
  }
}

impl Rule for A11yNestedInteractiveElementsRule {
  fn name(&self) -> &'static str {
    "a11y-nested-interactive-elements"
  }

  fn default_severity(&self) -> SeverityConfig {
    SeverityConfig::Severity(Severity::Error)
  }

  fn default_enabled(&self) -> bool {
    false
  }

  fn parser_options(&self) -> herb::ParserOptions {
    herb::ParserOptions {
      action_view_helpers: true,
      ..crate::rule::default_linter_parser_options()
    }
  }
}

impl ParserRule for A11yNestedInteractiveElementsRule {
  fn check(&self, result: &ParseResult, _context: &LintContext) -> Vec<UnboundOffense> {
    let mut visitor = NestedInteractiveElementsVisitor {
      rule_name: self.name(),
      offenses: Vec::new(),
      element_stack: ElementStack::new(),
    };

    visitor.visit_document_node(&result.value);

    visitor.offenses
  }
}
